plugins/producer_plugin/producer_plugin.cpp

```diff
@@ -2009,7 +2017,7 @@ bool producer_plugin_impl::process_unapplied_trxs( const fc::time_point& deadlin
       auto end_itr = (_pending_block_mode == pending_block_mode::producing) ?
                      _unapplied_transactions.unapplied_end()   : _unapplied_transactions.persisted_end();
       while( itr != end_itr ) {
-         if( deadline <= fc::time_point::now() ) {
+         if( deadline <= fc::time_point::now() || (_pending_block_mode == pending_block_mode::speculating && _received_block) ) {
```

Comment:
```suggestion
         if( deadline <= fc::time_point::now() || should_stop_speculating() ) {
```

Reply:
Added a `start_block_interrupted` method. This was much cleaner and handles all the cases. The only deadline still in play for speculative block is the block deadline passed to `push_transaction`. This really should be `fc::time_point::maximum()` for speculative (it will use max transaction time then), but I thought that a bit too much of a change for 3.1 & 3.2. I'll make that change for `main`.

Reply:
Nice!